File: dap/ws-server/dapClient.ts
import net from "net";
import { EventEmitter } from "events";
import path from "path";

// Used internally for requests that have not yet received a response.
export interface DAPMessage {
  seq: number;
  type: "request" | "response" | "event";
  command?: string;
  request_seq?: number;
  success?: boolean;
  body?: any;
  event?: string;
  arguments?: any;
}

const SEQ_UNASSIGNED = -1;

export class DAPClient extends EventEmitter {
  private socket: net.Socket;
  private buffer: string;
  private nextSeq: number;
  private pendingResponses: Map<number, DAPMessage>;
  private eventQueue: Map<string, DAPMessage[]>;

  // Optional callback that higher–level code (e.g. our WebSocket server)
  // can assign to broadcast DAP events.
  public broadcastFn?: (msg: DAPMessage) => void;

  constructor() {
    super();
    this.socket = new net.Socket();
    this.buffer = "";
    this.nextSeq = 1;
    this.pendingResponses = new Map();
    this.eventQueue = new Map();
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  // Connect to debugpy at host:port
  connect(host: string, port: number): Promise<void> {
    return new Promise((resolve, reject) => {
      this.socket.connect(port, host, () => {
        this.socket.on("data", (data: Buffer) => this.handleData(data));
        resolve();
      });
      this.socket.on("error", (err) => reject(err));
    });
  }

  // Send a DAP message with Content-Length header.
  sendMessage(message: DAPMessage): void {
    message.seq = this.nextSeq++;
    const json = JSON.stringify(message);
    const header = `Content-Length: ${Buffer.byteLength(json, "utf8")}\r\n\r\n`;
    const data = header + json;
    this.socket.write(data);

    console.log(
      `--> Sent (seq ${message.seq}, cmd: ${message.command}): ${json}`,
    );
  }

  // Wait for a specific response matching the requestSeq.
  waitForResponse(seq: number, timeout = 10000): Promise<DAPMessage> {
    return new Promise((resolve, reject) => {
      const start = Date.now();
      const interval = setInterval(() => {
        if (this.pendingResponses.has(seq)) {
          const msg = this.pendingResponses.get(seq)!;
          this.pendingResponses.delete(seq);
          clearInterval(interval);
          resolve(msg);
        } else if (Date.now() - start > timeout) {
          clearInterval(interval);
          reject(new Error(`Timeout waiting for response to seq ${seq}`));
        }
      }, 100);
    });
  }

  // Wait for a DAP event by name.
  waitForEvent(eventName: string, timeout = 10000): Promise<DAPMessage> {
    return new Promise((resolve, reject) => {
      let timer: NodeJS.Timeout | null = null;

      const checkQueue = () => {
        const existing = this.eventQueue.get(eventName);
        if (existing && existing.length > 0) {
          const msg = existing.shift()!;
          resolve(msg);
          if (timer) clearTimeout(timer);
        }
      };

      checkQueue();

      if (!timer) {
        const onEvent = () => {
          checkQueue();
        };
        this.on(eventName, onEvent);

        timer = setTimeout(() => {
          this.off(eventName, onEvent);
          reject(new Error(`Timeout waiting for event ${eventName}`));
        }, timeout);
      }
    });
  }

  // Main data handler for incoming data from the debugpy socket.
  private handleData(data: Buffer): void {
    this.buffer += data.toString("utf8");

    while (true) {
      const headerEnd = this.buffer.indexOf("\r\n\r\n");
      if (headerEnd === -1) break;

      const header = this.buffer.slice(0, headerEnd);
      const match = header.match(/Content-Length:\s*(\d+)/);
      if (!match) {
        this.emit("error", new Error("No Content-Length header found"));
        return;
      }
      const length = parseInt(match[1], 10);
      const totalMessageLength = headerEnd + 4 + length;
      if (this.buffer.length < totalMessageLength) break;

      const body = this.buffer.slice(headerEnd + 4, totalMessageLength);
      this.buffer = this.buffer.slice(totalMessageLength);

      let msg: DAPMessage;
      try {
        msg = JSON.parse(body);
      } catch (e) {
        console.error("Error parsing JSON message", e);
        continue;
      }

      // If it is a response, store it in pendingResponses.
      if (msg.type === "response" && msg.request_seq) {
        this.pendingResponses.set(msg.request_seq, msg);
      }

      // For events, enqueue them and
      if (msg.type === "event" && msg.event) {
        if (!this.eventQueue.has(msg.event)) {
          this.eventQueue.set(msg.event, []);
        }
        this.eventQueue.get(msg.event)!.push(msg);
        this.emit(msg.event, msg);

        // Call the broadcast callback if it is set.
        if (this.broadcastFn) {
          this.broadcastFn(msg);
        }
      }

      // Emit the generic message for logging and further processing.
      this.emit("message", msg);
      console.log("<-- Received:", msg);
    }
  }

  // HIGHER-LEVEL REQUEST METHODS

  async initialize(): Promise<DAPMessage> {
    const initSeq = this.nextSeq;
    const req: DAPMessage = {
      seq: SEQ_UNASSIGNED,
      type: "request",
      command: "initialize",
      arguments: {
        adapterID: "python",
        clientID: "dap_test_client",
        clientName: "DAP Test",
        linesStartAt1: true,
        columnsStartAt1: true,
        pathFormat: "path",
        supportsVariableType: true,
        supportsEvaluateForHovers: true,
      },
    };
    this.sendMessage(req);
    return this.waitForResponse(initSeq);
  }

  async attach(host: string, port: number): Promise<void> {
    const req: DAPMessage = {
      seq: SEQ_UNASSIGNED,
      type: "request",
      command: "attach",
      arguments: { host, port },
    };
    this.sendMessage(req);
    await this.sleep(200);
    await this.waitForEvent("initialized");
  }

  async setBreakpoints(
    filePath: string,
    breakpoints: Array<{ line: number }>,
  ): Promise<DAPMessage> {
    const bpSeq = this.nextSeq;
    const req: DAPMessage = {
      seq: SEQ_UNASSIGNED,
      type: "request",
      command: "setBreakpoints",
      arguments: {
        source: {
          path: filePath,
          name: path.basename(filePath),
        },
        breakpoints,
        sourceModified: false,
      },
    };
    this.sendMessage(req);
    return this.waitForResponse(bpSeq);
  }

  async configurationDone(): Promise<DAPMessage> {
    const confSeq = this.nextSeq;
    const req: DAPMessage = {
      seq: SEQ_UNASSIGNED,
      type: "request",
      command: "configurationDone",
      arguments: {},
    };
    this.sendMessage(req);
    return this.waitForResponse(confSeq);
  }

  async tryGetAttachResponse(
    attachSeq: number,
    timeout = 1000,
  ): Promise<DAPMessage | null> {
    try {
      const resp = await this.waitForResponse(attachSeq, timeout);
      return resp;
    } catch (err) {
      console.log("No attach response received", err);
      return null;
    }
  }

  async continue(threadId: number): Promise<DAPMessage> {
    const contSeq = this.nextSeq;
    const req: DAPMessage = {
      seq: SEQ_UNASSIGNED,
      type: "request",
      command: "continue",
      arguments: { threadId },
    };
    this.sendMessage(req);
    return this.waitForResponse(contSeq);
  }

  async stackTrace(
    threadId: number,
    startFrame = 0,
    levels = 1,
  ): Promise<DAPMessage> {
    const stSeq = this.nextSeq;
    const req: DAPMessage = {
      seq: SEQ_UNASSIGNED,
      type: "request",
      command: "stackTrace",
      arguments: { threadId, startFrame, levels },
    };
    this.sendMessage(req);
    return this.waitForResponse(stSeq);
  }

  async evaluate(expression: string, frameId?: number): Promise<DAPMessage> {
    const evalSeq = this.nextSeq;
    const args: any = { expression, context: "hover" };
    if (frameId) {
      args.frameId = frameId;
    }
    const req: DAPMessage = {
      seq: SEQ_UNASSIGNED,
      type: "request",
      command: "evaluate",
      arguments: args,
    };
    this.sendMessage(req);
    return this.waitForResponse(evalSeq);
  }

  close(): void {
    this.socket.end();
  }
}

File: dap/rust_tests/src/dap_client.rs
use serde::de::{self, Visitor};
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::collections::HashMap;
use std::fmt;
use std::io::{BufReader, ErrorKind, Read, Write};
use std::net::TcpStream;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

#[derive(Debug, Clone, PartialEq)]
pub enum MessageType {
    Request,
    Response,
    Event,
}

// Manually implement Serialize so that the variants are always lower-case.
impl Serialize for MessageType {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let s = match *self {
            MessageType::Request => "request",
            MessageType::Response => "response",
            MessageType::Event => "event",
        };
        serializer.serialize_str(s)
    }
}

struct MessageTypeVisitor;

impl<'de> Visitor<'de> for MessageTypeVisitor {
    type Value = MessageType;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("a string representing a message type")
    }

    fn visit_str<E>(self, value: &str) -> Result<MessageType, E>
    where
        E: de::Error,
    {
        match value {
            "request" => Ok(MessageType::Request),
            "response" => Ok(MessageType::Response),
            "event" => Ok(MessageType::Event),
            _ => Err(de::Error::unknown_variant(
                value,
                &["request", "response", "event"],
            )),
        }
    }
}

impl<'de> Deserialize<'de> for MessageType {
    fn deserialize<D>(deserializer: D) -> Result<MessageType, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_str(MessageTypeVisitor)
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct DAPMessage {
    pub seq: i32,
    #[serde(rename = "type")]
    pub message_type: MessageType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub command: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request_seq: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub success: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub event: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub arguments: Option<serde_json::Value>,
}

pub struct DAPClient {
    writer: Option<Arc<Mutex<TcpStream>>>,
    // We keep an Arc of the BufReader ensure the receiver thread is the only one that reads.
    reader: Option<Arc<Mutex<BufReader<TcpStream>>>>,
    next_seq: Arc<Mutex<i32>>,
    // Responses are added by the receiver thread after reading from the socket.
    responses: Arc<Mutex<HashMap<i32, DAPMessage>>>,
    events: Arc<Mutex<HashMap<String, Vec<DAPMessage>>>>,
    receiver_handle: Option<thread::JoinHandle<()>>,
}

impl DAPClient {
    pub fn new() -> Self {
        Self {
            writer: None,
            reader: None,
            next_seq: Arc::new(Mutex::new(1)),
            responses: Arc::new(Mutex::new(HashMap::new())),
            events: Arc::new(Mutex::new(HashMap::new())),
            receiver_handle: None,
        }
    }

    pub fn connect(&mut self, host: &str, port: u16) -> std::io::Result<()> {
        let stream = TcpStream::connect((host, port))?;
        // Note: We clone the stream so that writer and reader can be used on separate handles.
        self.writer = Some(Arc::new(Mutex::new(stream.try_clone()?)));
        self.reader = Some(Arc::new(Mutex::new(BufReader::new(stream))));
        Ok(())
    }

    /// Sends a DAP message. Returns the assigned sequence number.
    pub fn send_message(&self, mut message: DAPMessage) -> std::io::Result<i32> {
        let seq = {
            let mut seq_lock = self.next_seq.lock().unwrap();
            let current = *seq_lock;
            *seq_lock += 1;
            current
        };

        message.seq = seq;
        let json = serde_json::to_string(&message)?;
        let header = format!("Content-Length: {}\r\n\r\n", json.len());
        println!(
            "--> Sending message (seq={}):\nHeader: {}\nPayload: {}",
            seq, header, json
        );

        let mut writer = self.writer.as_ref().unwrap().lock().unwrap();
        writer.write_all(header.as_bytes())?;
        writer.write_all(json.as_bytes())?;
        writer.flush()?;
        Ok(seq)
    }

    /// Starts a background receiver thread which is the only code that reads the socket.
    pub fn start_receiver(&mut self) {
        // Clone the Arcs needed by the receiver thread.
        let reader_arc = Arc::clone(self.reader.as_ref().unwrap());
        let responses_arc = Arc::clone(&self.responses);
        let events_arc = Arc::clone(&self.events);

        self.receiver_handle = Some(thread::spawn(move || {
            loop {
                // Read header until we find the "\r\n\r\n" sequence.
                let header = {
                    let mut reader = reader_arc.lock().unwrap();
                    let mut header_bytes = Vec::new();
                    // Read one byte at a time until the header terminator is found.
                    loop {
                        let mut buf = [0u8; 1];
                        match reader.read_exact(&mut buf) {
                            Ok(()) => {
                                header_bytes.push(buf[0]);
                                if header_bytes.ends_with(b"\r\n\r\n") {
                                    break;
                                }
                            }
                            Err(ref e) if e.kind() == ErrorKind::UnexpectedEof => {
                                // Connection closed.
                                return;
                            }
                            Err(e) => {
                                eprintln!("Error reading header: {}", e);
                                return;
                            }
                        }
                    }
                    String::from_utf8_lossy(&header_bytes).to_string()
                };

                // Parse Content-Length from header.
                let content_length = header
                    .lines()
                    .find(|line| line.to_lowercase().starts_with("content-length:"))
                    .and_then(|line| line[15..].trim().parse::<usize>().ok());

                if let Some(len) = content_length {
                    // Now, read the body.
                    let mut body_bytes = vec![0; len];
                    {
                        let mut reader = reader_arc.lock().unwrap();
                        if let Err(e) = reader.read_exact(&mut body_bytes) {
                            eprintln!("Error reading body: {}", e);
                            return;
                        }
                    }
                    let message_str = match String::from_utf8(body_bytes) {
                        Ok(s) => s,
                        Err(e) => {
                            eprintln!("Invalid UTF-8 body: {}", e);
                            continue;
                        }
                    };
                    println!("<-- Received: {}", message_str);
                    match serde_json::from_str::<DAPMessage>(&message_str) {
                        Ok(msg) => match msg.message_type {
                            MessageType::Response => {
                                if let Some(req_seq) = msg.request_seq {
                                    responses_arc.lock().unwrap().insert(req_seq, msg);
                                }
                            }
                            MessageType::Event => {
                                if let Some(evt) = &msg.event {
                                    events_arc
                                        .lock()
                                        .unwrap()
                                        .entry(evt.clone())
                                        .or_insert_with(Vec::new)
                                        .push(msg);
                                }
                            }
                            _ => {}
                        },
                        Err(e) => {
                            eprintln!("Error parsing message: {}", e);
                        }
                    }
                } else {
                    eprintln!("No Content-Length found in header: {}", header);
                }
                // Don't busy‐spin.
                thread::sleep(Duration::from_millis(10));
            }
        }));
    }

    /// Waits for an event (by name) until the timeout (in seconds) expires.
    pub fn wait_for_event(&self, name: &str, timeout_secs: f64) -> Option<DAPMessage> {
        let start = Instant::now();
        while start.elapsed().as_secs_f64() < timeout_secs {
            if let Some(mut events) = self.events.lock().ok() {
                if let Some(list) = events.get_mut(name) {
                    if !list.is_empty() {
                        return Some(list.remove(0));
                    }
                }
            }
            thread::sleep(Duration::from_millis(50));
        }
        None
    }

    /// Waits asynchronously for a response with the given sequence number.
    pub async fn wait_for_response(&self, seq: i32, timeout_secs: f64) -> Option<DAPMessage> {
        let start = Instant::now();
        while start.elapsed().as_secs_f64() < timeout_secs {
            if let Some(resp) = self.responses.lock().unwrap().remove(&seq) {
                return Some(resp);
            }
            tokio::time::sleep(Duration::from_millis(50)).await;
        }
        None
    }

    // Now the DAP commands use only send_message and wait_for_response.
    pub async fn initialize(&self) -> Result<DAPMessage, Box<dyn std::error::Error>> {
        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("initialize".to_string()),
            request_seq: None,
            success: None,
            arguments: Some(serde_json::json!({
                "adapterID": "python",
                "clientID": "dap_test_client",
                "clientName": "DAP Test",
                "linesStartAt1": true,
                "columnsStartAt1": true,
                "pathFormat": "path",
                "supportsVariableType": true,
                "supportsEvaluateForHovers": true,
            })),
            body: None,
            event: None,
        };

        let seq = self.send_message(req)?;
        if let Some(response) = self.wait_for_response(seq, 10.0).await {
            Ok(response)
        } else {
            Err("Timeout waiting for initialize response".into())
        }
    }

    pub fn attach(&self, host: &str, port: u16) -> std::io::Result<()> {
        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("attach".to_string()),
            request_seq: None,
            success: None,
            body: None,
            event: None,
            arguments: Some(serde_json::json!({
                "host": host,
                "port": port
            })),
        };
        // Send the message and do not wait for a response.
        self.send_message(req)?;
        Ok(())
    }

    pub async fn configuration_done(&self) -> Result<DAPMessage, Box<dyn std::error::Error>> {
        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("configurationDone".to_string()),
            request_seq: None,
            success: None,
            body: None,
            event: None,
            arguments: None,
        };
        let seq = self.send_message(req)?;
        if let Some(response) = self.wait_for_response(seq, 10.0).await {
            Ok(response)
        } else {
            Err("Timeout waiting for configurationDone response".into())
        }
    }

    pub async fn set_breakpoints(
        &self,
        file_path: &str,
        line_numbers: &[i32],
    ) -> Result<DAPMessage, Box<dyn std::error::Error>> {
        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("setBreakpoints".to_string()),
            request_seq: None,
            success: None,
            body: None,
            event: None,
            arguments: Some(serde_json::json!({
                "source": {
                    "path": file_path,
                    "name": std::path::Path::new(file_path).file_name().unwrap().to_str().unwrap()
                },
                "breakpoints": line_numbers.iter().map(|&line| serde_json::json!({ "line": line })).collect::<Vec<_>>(),
                "sourceModified": false
            })),
        };

        let seq = self.send_message(req)?;
        if let Some(response) = self.wait_for_response(seq, 10.0).await {
            Ok(response)
        } else {
            Err("Timeout waiting for setBreakpoints response".into())
        }
    }

    pub async fn stack_trace(
        &self,
        thread_id: i32,
    ) -> Result<DAPMessage, Box<dyn std::error::Error>> {
        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("stackTrace".to_string()),
            request_seq: None,
            success: None,
            body: None,
            event: None,
            arguments: Some(serde_json::json!({
                "threadId": thread_id,
                "startFrame": 0,
                "levels": 1
            })),
        };

        let seq = self.send_message(req)?;
        if let Some(response) = self.wait_for_response(seq, 10.0).await {
            Ok(response)
        } else {
            Err("Timeout waiting for stackTrace response".into())
        }
    }

    pub async fn evaluate(
        &self,
        expression: &str,
        frame_id: Option<i32>,
    ) -> Result<DAPMessage, Box<dyn std::error::Error>> {
        let mut args = serde_json::json!({
            "expression": expression,
            "context": "hover"
        });

        if let Some(id) = frame_id {
            if let serde_json::Value::Object(ref mut map) = args {
                map.insert("frameId".to_string(), serde_json::json!(id));
            }
        }

        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("evaluate".to_string()),
            request_seq: None,
            success: None,
            body: None,
            event: None,
            arguments: Some(args),
        };

        let seq = self.send_message(req)?;
        if let Some(response) = self.wait_for_response(seq, 10.0).await {
            Ok(response)
        } else {
            Err("Timeout waiting for evaluate response".into())
        }
    }

    pub async fn continue_execution(
        &self,
        thread_id: i32,
    ) -> Result<DAPMessage, Box<dyn std::error::Error>> {
        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("continue".to_string()),
            request_seq: None,
            success: None,
            body: None,
            event: None,
            arguments: Some(serde_json::json!({
                "threadId": thread_id
            })),
        };

        let seq = self.send_message(req)?;
        if let Some(response) = self.wait_for_response(seq, 10.0).await {
            Ok(response)
        } else {
            Err("Timeout waiting for continue response".into())
        }
    }

    pub async fn step_in(
        &self,
        thread_id: i32,
        target_id: Option<i32>,
        granularity: &str,
    ) -> Result<DAPMessage, Box<dyn std::error::Error>> {
        let mut args = serde_json::json!({
            "threadId": thread_id,
            "granularity": granularity
        });

        if let Some(id) = target_id {
            if let serde_json::Value::Object(ref mut map) = args {
                map.insert("targetId".to_string(), serde_json::json!(id));
            }
        }

        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("stepIn".to_string()),
            request_seq: None,
            success: None,
            body: None,
            event: None,
            arguments: Some(args),
        };

        let seq = self.send_message(req)?;
        if let Some(response) = self.wait_for_response(seq, 10.0).await {
            Ok(response)
        } else {
            Err("Timeout waiting for stepIn response".into())
        }
    }

    pub async fn next(&self, thread_id: i32) -> Result<DAPMessage, Box<dyn std::error::Error>> {
        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("next".to_string()),
            request_seq: None,
            success: None,
            body: None,
            event: None,
            arguments: Some(serde_json::json!({
                "threadId": thread_id
            })),
        };

        let seq = self.send_message(req)?;
        if let Some(response) = self.wait_for_response(seq, 10.0).await {
            Ok(response)
        } else {
            Err("Timeout waiting for next response".into())
        }
    }

    pub async fn step_out(
        &self,
        thread_id: i32,
        granularity: &str,
    ) -> Result<DAPMessage, Box<dyn std::error::Error>> {
        let req = DAPMessage {
            seq: -1,
            message_type: MessageType::Request,
            command: Some("stepOut".to_string()),
            request_seq: None,
            success: None,
            body: None,
            event: None,
            arguments: Some(serde_json::json!({
                "threadId": thread_id,
                "granularity": granularity
            })),
        };

        let seq = self.send_message(req)?;
        if let Some(response) = self.wait_for_response(seq, 10.0).await {
            Ok(response)
        } else {
            Err("Timeout waiting for stepOut response".into())
        }
    }
}
